sql/range.go

```diff
@@ -111,64 +200,116 @@ func (rang Range) IsSupersetOf(otherRange Range) (bool, error) {
 	return otherRange.IsSubsetOf(rang)
 }
 
-// Equals returns whether the calling RangeColumn is equal to the given RangeColumn.
-func (rc RangeColumn) Equals(other RangeColumn) (bool, error) {
-	if len(rc) != len(other) {
+// Overlaps returns whether the calling Range and given Range have overlapping values, which would result in the same
+// values being returned from some subset of both ranges.
+func (rang Range) Overlaps(otherRange Range) (bool, error) {
+	if len(rang) != len(otherRange) {
 		return false, nil
 	}
-	for i := range rc {
-		if ok, err := rc[i].Equals(other[i]); err != nil || !ok {
+	for i := range rang {
+		_, ok, err := rang[i].Overlaps(otherRange[i])
+		if err != nil || !ok {
 			return false, err
 		}
 	}
 	return true, nil
 }
 
-// TryIntersect attempts to intersect the calling RangeColumn with the given RangeColumn. If the intersection fails or
-// results in an empty RangeColumn then nil and false are returned.
-func (rc RangeColumn) TryIntersect(other RangeColumn) (RangeColumn, bool, error) {
-	var newRangeColumn []RangeColumnExpr
-	var err error
-	for _, rr := range rc {
-		for _, or := range other {
-			newRange, ok, err := rr.TryIntersect(or)
-			if err != nil {
-				return nil, false, err
-			}
-			if ok {
-				newRangeColumn = append(newRangeColumn, newRange)
-			}
-		}
+// RemoveOverlap removes any overlap that the given Range may have with the calling Range. If the two ranges do not
+// overlap then they're both returned. If one is a subset of the other/may be unioned then only one Range is returned.
+// Otherwise, this returns a set of ranges that do not overlap with each other.
+//
+// If the two ranges do not overlap then false is returned, otherwise returns true.
+func (rang Range) RemoveOverlap(otherRange Range) (RangeCollection, bool, error) {
```

Comment:
Not so. `TryMerge` attempts to merge the two ranges so that only one range is returned. For most cases that have overlapping ranges, only a small section may be merged between the two, but other parts do not overlap at all and therefore cannot be merged.

Reply:
This is what I don't get. I'm picturing this as ranges on a number line, so if two of them overlap, they can be merged together. That's obviously too simplistic, but I don't understand why. Just because it's N-dimensional? Think this needs some comments with concrete examples so that people who come after you can understand what's going on.